### DISCUSSION TITLE
Meeting screen, enhancement for switching troops between heroes with the big yellow arrows [Done]

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#4327] [Ideas]
I propose to add a central clickable area that allows to switch the complete armies between the two heroes.
The top and bottom clickable areas allow to transfer as many troops as possible between the armies of the two heroes, as it is currently the case.

A central clickable area is also added for artifacts that allows to switch all artifacts from each hero to the other hero.

To indicate to the player what action is going to be performed, when the mouse cursor is positioned over an area (before clicking), the color of the arrow/arrows is changed (for example to green).
For the top and bottom areas, only the respective arrow changes color, for the middle area both arrows change color.
The arrows can also be slightly distanced to increase the surface of the central area.

In fheroes2:

![Meeting screen switch area 01](https://user-images.githubusercontent.com/43583994/136184968-209f862d-82a0-4c54-be78-cca3b0dc7504.png)

**Comment by Branikolog** [comment]
We definitely need fast exchange option here. But I prefer adding one more button rather than click empty area even with highlighting effects.
What about adding there a new button in similar design? 
Something like:
![Button Side](https://user-images.githubusercontent.com/55348946/136370845-7bac89ba-c2ce-4e1e-890b-03fa143ea9ad.png)
I think, we have enough free space here for one more useful button.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

> We definitely need fast exchange option here. But I prefer adding one more button rather than click empty area even with highlighting effects.
> What about adding there a new button in similar design?
> Something like:
> ![Button Side](https://user-images.githubusercontent.com/55348946/136370845-7bac89ba-c2ce-4e1e-890b-03fa143ea9ad.png)
> I think, we have enough free space here for one more useful button.

I made the proposal keeping in mind not to change the current button (at least at first).
Now, if another button can be added that's even better. :)
Your button proposal seems to fit perfectly.

**Reply by Branikolog** [reply]
@LeHerosInconnu 
I definitely don't like clicking empty area. Also, player could even don't know it being clickable, until he points that place. So a separate button should be the only solution for this feature.
This button can be simply combined from the existing ones, I believe. The only difficult issue is to place it properly on he screen. I suppose that we cannot easily move current buttons.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> @LeHerosInconnu I definitely don't like clicking empty area. Also, player could even don't know it being clickable, until he points that place. So a separate button should be the only solution for this feature.
> This button can be simply combined from the existing ones, I believe. The only difficult issue is to place it properly on he screen. I suppose that we cannot easily move current buttons.

Good point.
I think it should not be difficult to insert your double arrow button between the two other buttons already present.
After all, it's just a matter of positioning the graphics (and possibly cutting the ones already in use).

**Reply by Branikolog** [reply]
I'd prefer moving the existing ones a bit, as there're not enough free space between them, and placing new button higher or lower won't look good.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> I'd prefer moving the existing ones a bit, as there're not enough free space between them, and placing new button higher or lower won't look good.

I was thinking of spacing the two arrows already present and inserting the new button between these arrows at equal distance. :)
If this is not possible, the new button would be better positioned visually below the arrows already present, and with similar spacing, the whole group of buttons then centered vertically with respect to the inventories.

**Reply by zenseii** [reply]
Hi @LeHerosInconnu and @Branikolog!
From what I understand it would be best to have this new button in between the two original arrow buttons? This should be easy to arrange.

**Reply by LeHerosInconnu** [reply]
Hello @zenseii,

> Hi @LeHerosInconnu and @Branikolog! From what I understand it would be best to have this new button in between the two original arrow buttons?

Yes, between the two yellow arrows already present would be better.

> This should be easy to arrange.

Nice. :)

**Reply by Branikolog** [reply]
Hi, @zenseii !
It would be great, of new button would appear here. Players are complaining, we don't have swap function here.
To my mind, more space between move buttons (which will obviously be taken by swap button) would prevent player from accidental uniting armies on wrong hero: swap button close to each move button will lead to simple swapping armies, which can be reverted easily.

**Reply by zenseii** [reply]
> Hi, @zenseii ! It would be great, of new button would appear here. Players are complaining, we don't have swap function here. To my mind, more space between move buttons (which will obviously be taken by swap button) would prevent player from accidental uniting armies on wrong hero: swap button close to each move button will lead to simple swapping armies, which can be reverted easily.

Hi @Branikolog! Yes, I agree that the current move button is borderline useless and many will slowly realize that it oftentimes is better to just manually move every single troop.

You bring up a good point about not having these buttons too close to each other to avoid accidental clicks. I think we should keep this in mind, especially for when a smartphone port happens because most people have fat thumbs :P I did have a thought about putting single arrows above or below the armybars (: the actual line of troops) in the heroes meeting screen, but I was hesistant to suggest it since I felt it would change the window drastically from its original design.

**Comment by zenseii** [comment]
It would be something like this:
![image](https://user-images.githubusercontent.com/12501091/172577864-e7ddb996-218e-40ec-948c-9d1a19b04eaa.png)

Honestly you could even just retain the original two arrows and just have that be a single button that does the swap. I think the way they point in opposite directions makes sense to perform a swap, which is also what I think many players who press it think it does.

**Reply by zenseii** [reply]
Here's the example using the original arrows:
![image](https://user-images.githubusercontent.com/12501091/172578894-78819849-bb0d-476a-b847-eaefe6ed13a5.png)

I think that one is better an simpler, and doesn't change the original esthetic too much.

**Reply by LeHerosInconnu** [reply]
From an ergonomic point of view, your proposal forces the player to go miles with the mouse, I would prefer bigger buttons but located near each other. :)

**Reply by zenseii** [reply]
It is true that the distance between the swap and the move button increases, but do you ever enter this screen with the intent to do both a swap and a move? You would already know if you want to do a swap or a move, no? The point is that there would never be a need to have these buttons close to each other.

Having them distanced does in fact avoid the problem of misclicks.

**Reply by zenseii** [reply]
Another issue with having all these arrows in the center is that if someone brings up the idea that we should have swap and move for artifacts too, then this central column becomes quite crowded. I'll try to have an image up.

**Reply by LeHerosInconnu** [reply]
Never underestimate the uses of the interface by the users, even if they are bad. :)

I thought this was going to be done for artifacts too, why do things by halves. :rofl:

**Comment by zenseii** [comment]
Please try to look aside from the fact that they are not properly center-aligned.
![image](https://user-images.githubusercontent.com/12501091/172580639-d2b19160-4ed8-41e4-a756-3fe047ea62a3.png)

**Reply by LeHerosInconnu** [reply]
> Please try to look aside from the fact that they are not properly aligned.

That's what I see.
You are forgiven. :D

Let me take some time.

**Reply by LeHerosInconnu** [reply]
For the PC (or other) version, with a mouse, it is not a problem if the buttons are much closer together.
For the touch interface version, it would be possible to make it so that when the user presses the button area, the different buttons are zoomed (bigger buttons are displayed) to allow the user a more comfortable use.
And this could also be an option for the PC version for users who often make missclicks and also in the case of a high game resolution.

**Reply by Branikolog** [reply]
Well, both of implementations have its pros and cons.
In my personal point of view, all buttons in the middle look a little crowded. Also, buttons look almost the same: 3 yellow lines and it's rather hard to distinguish, what arrow provides necessary action.
Move buttons being moved make 100% clear impression, what do they for. They look a bit unusual, but acceptable.
I only prefer them being under the boxes it interacts. (For artifacts we need to place such arrows in the bottom of the screen too).

**Reply by LeHerosInconnu** [reply]
Hello @zenseii and @Branikolog,

It is also possible to use different buttons.
Here is an example with buttons used for the "extended inventories" of my mod, explained [here](https://github.com/ihhub/fheroes2/discussions/2995):

Heroes meeting BAGS screen:
![Heroes Meeting BAGS Screen enhanced and Switch Add and top buttons](https://user-images.githubusercontent.com/43583994/111880254-9c05bf00-89aa-11eb-96c4-37022d8d8bea.png)

This would look like this:

![Meeting screen 10](https://user-images.githubusercontent.com/43583994/172812802-9e42320d-6285-49c4-b264-aa8bb77f2de7.png)

And with the buttons zoomed in as explained above (the buttons could be zoomed in even larger, and positioned elsewhere):

![Meeting screen zoom 10](https://user-images.githubusercontent.com/43583994/172812869-37ea2bab-93c2-4381-8690-9d6496f7feb8.png)

Images are for explanation only.

**Comment by LeHerosInconnu** [comment]
This has been implemented.
I close the discussion.
